integration/kotlinx-coroutines-guava/src/ListenableFuture.kt

```diff
@@ -5,31 +5,51 @@
 package kotlinx.coroutines.guava
 
 import com.google.common.util.concurrent.*
-import kotlinx.coroutines.*
 import java.util.concurrent.*
-import java.util.concurrent.CancellationException
 import kotlin.coroutines.*
+import kotlinx.coroutines.*
+import java.lang.IllegalStateException
 
 /**
- * Starts new coroutine and returns its results an an implementation of [ListenableFuture].
- * The running coroutine is cancelled when the resulting future is cancelled or otherwise completed.
+ * Starts [block] in a new coroutine and returns a [ListenableFuture] pointing to its result.
+ *
+ * The coroutine is immediately started. Passing [CoroutineStart.LAZY] to [start] throws
+ * [IllegalArgumentException], because Futures don't have a way to start lazily.
+ *
+ * The created coroutine is cancelled when the resulting future completes successfully, fails, or
+ * is cancelled.
+ *
+ * `CoroutineContext` is inherited from this [CoroutineScope]. Additional context elements can be
+ * added/overlaid by passing [context].
  *
- * Coroutine context is inherited from a [CoroutineScope], additional context elements can be specified with [context] argument.
- * If the context does not have any dispatcher nor any other [ContinuationInterceptor], then [Dispatchers.Default] is used.
- * The parent job is inherited from a [CoroutineScope] as well, but it can also be overridden
- * with corresponding [coroutineContext] element.
+ * If the context does not have a [CoroutineDispatcher], nor any other [ContinuationInterceptor]
+ * member, [Dispatchers.Default] is used.
  *
- * By default, the coroutine is immediately scheduled for execution.
- * Other options can be specified via `start` parameter. See [CoroutineStart] for details.
- * A value of [CoroutineStart.LAZY] is not supported
- * (since `ListenableFuture` framework does not provide the corresponding capability) and
- * produces [IllegalArgumentException].
+ * The parent job is inherited from this [CoroutineScope], and can be overridden by passing
+ * a [Job] in [context].
  *
- * See [newCoroutineContext][CoroutineScope.newCoroutineContext] for a description of debugging facilities that are available for newly created coroutine.
+ * See [newCoroutineContext][CoroutineScope.newCoroutineContext] for a description of debugging
+ * facilities.
  *
- * @param context additional to [CoroutineScope.coroutineContext] context of the coroutine.
+ * Note that the error semantics of [future] are _subtly different_ than [asListenableFuture].
+ * By calling [future], any error that occurs after successfully cancelling the returned
+ * [ListenableFuture] will be passed to the [CoroutineExceptionHandler] from the context. The
+ * contract of [Future] does not permit it to return an error after it is successfully cancelled.
+ *
+ * By calling [asListenableFuture] on a [Deferred], any error that occurs after successfully
```

Comment:
This paragraph is not necessary for the public API, it's better to move to implementation

Reply:
Moved to `CancellationToCoroutine` class doc and crosslinked.

Reply:
Move to `ListenableFutureCoroutine` after merging `CancellationToCoroutine`.